### PR TITLE
#520: Move common headers matcher logic to AbstractHmHeader

### PR DESCRIPTION
**Pull request by Happy-Neko**
Fix for #520.
* Move common headers matcher logic to `AbstractHmHeader`
* Unify error messages between `HmRqHeader` and `HsRqHeader`

### CODE REVIEW

**Comment by Happy-Neko** [comment]
@davvd please proceed with code review.

**Comment by davvd** [comment]
@happy-neko Thanks, I will ask someone to do a code review here, thanks for your pull request

**Comment by davvd** [comment]
@longtimeago this pull request is for you,please review

File: src/main/java/org/takes/facets/hamcrest/AbstractHmHeader.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    @Override
+    public final boolean matchesSafely(final T item) {
+        try {
+            final Iterator<String> headers = headers(item).iterator();
+            if (headers.hasNext()) {
+                headers.next();
+            }
+            final List<String> values = new ArrayList<String>(0);
```
Comment (longtimeago) [inline src/main/java/org/takes/facets/hamcrest/AbstractHmHeader.java]: @Happy-Neko Variable type can be relaxed to `Collection`

File: src/main/java/org/takes/facets/hamcrest/HmRqHeader.java [hunk 2]
```diff
@@ -24,194 +24,53 @@
...
-import java.util.Map;
-import org.hamcrest.Description;
 import org.hamcrest.Matcher;
-import org.hamcrest.TypeSafeMatcher;
 import org.takes.Request;
-import org.takes.misc.EntryImpl;
-import org.takes.rq.RqHeaders;
 
 /**
  * Response Header Matcher.
```
Comment (longtimeago) [inline src/main/java/org/takes/facets/hamcrest/HmRqHeader.java]: @Happy-Neko This is Request Header Matcher, not a Response

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 3]
```diff
@@ -134,7 +141,32 @@ public void testsMultipleHeadersAvailable() throws Exception {
...
+            )
+        );
+    }
+
+    /**
+     * Checks is mismatch message is readable.
+     */
+    @Test
+    public void testMismatchMessage() {
+        final String expected = new StringBuilder()
```
Comment (longtimeago) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @Happy-Neko This is a one-time variable. Please, inline it

**Comment by longtimeago** [comment]
@Happy-Neko please, see several minor comments above. Thanks!

**Comment by Happy-Neko** [comment]
@longtimeago fix committed, thanks.

**Comment by longtimeago** [comment]
@Happy-Neko Thanks, now it's good!

**Comment by longtimeago** [comment]
@rultor merge

File: src/main/java/org/takes/facets/hamcrest/AbstractHmHeader.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+ * <p>This "matcher" tests given item headers.
+ * <p>The class is immutable and thread-safe.
+ *
+ * @param <T> Item type. Should be able to return own headers
+ * @author Eugene Kondrashev (eugene.kondrashev@gmail.com)
+ * @author I. Sokolov (happy.neko@gmail.com)
+ * @author Andrey Eliseev (aeg.exper0@gmail.com)
+ * @version $Id$
+ */
+public abstract class AbstractHmHeader<T> extends TypeSafeMatcher<T> {
```
Comment (yegor256) [inline src/main/java/org/takes/facets/hamcrest/AbstractHmHeader.java]: @Happy-Neko why public?

**Comment by yegor256** [comment]
@Happy-Neko see above

**Comment by Happy-Neko** [comment]
@yegor256 changes committed, thanks.

**Comment by yegor256** [comment]
@rultor merge

**Comment by longtimeago** [comment]
@rultor deploy

**Comment by yegor256** [comment]
@rultor deploy

**Comment by davvd** [comment]
@elenavolokhova please, help us to evaluate quality here, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thank you ;)

**Comment by davvd** [comment]
@longtimeago paied 10 mins to @elenavolokhova for QA review (payment ID is `75801733`); Thanks so much! Your account was topped up for **22 mins** (transaction ID is `75801775`, the task took 48 hours and 52 mins); review comments (c=7) added as a bonus; +22 to your rating, your total score is [+3545](http://www.netbout.com/b/35024?open=rating)

**Pull request merged**